klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,194 @@
+    When directories are created, moved or deleted, we will be notified here.
+    We try and update the klaus wsgi app with modifications to git repos as
+    they happen.
+    """
+    def __init__(self, klausapp, *args, **kwargs):
+        self._klausapp = klausapp
+        return super(KlausFSEventHandler, self).__init__(*args, **kwargs)
+
+    def on_modified(self, event):
+        if not event.is_directory:
+            return
+        self._klausapp.remove_repo(event.src_path)
+        self._klausapp.add_repo(event.src_path)
+
+    def on_created(self, event):
+        # We're only interested in directories; A git repo is always a directory
+        if not event.is_directory:
+            return
+        logger.debug("Received directory creation event for '%s'", event.src_path)
+        self._klausapp.add_repo(event.src_path)
+
+    def on_deleted(self, event):
+        if not event.is_directory:
+            return
+        logger.debug("Received directory deletion event for '%s'", event.src_path)
+        self._klausapp.remove_repo(event.src_path)
+
+    def on_moved(self, event):
+        """
+        Moved is not the same as deleted, unless the moved dir is no longer a
+        subset of any of the watched trees. In that case we need to consider it
+        a deletion, and remove it from the list of watched repos; otherwise we
+        need to update its path. This is a potentially expensive operation,
+        but shouldn't affect performance in the general case, as moving a
+        repository is a rare event.
+        """
+        logger.debug(
+            "Received repository move event:'%s'->'%s'",
+            event.src_path,
+            event.dest_path
+        )
+        # first unregister old repository
+        if not self._klausapp.remove_repo(event.src_path):
+            logger.debug(
+                "ignored non-repo move, %s -> %s",
+                event.src_path, event.dest_path
+            )
+            return
+
+        # then see if the new destination is part of the klaus watchlist
+        for directory in self._klausapp.basedirs:
+            if is_subdirectory(directory, event.dest_path):
+                logger.info(
+                    "'%s' -> '%s'; updating references",
+                    event.src_path,
+                    event.dest_path
+                )
+                self._klausapp.add_repo(event.dest_path)
+                break
+        else:
+            logger.info(
+                "%s was moved out of KLAUS_REPOS to %s, and is now ignored",
+                event.src_path,
+                event.dest_path
+            )
+
+
+class KlausAutoDiscover(Klaus):
+    """
+    Instead of the normal Klaus behaviour of receiving a list of git
+    repositories as `os.environ['KLAUS_REPOS']`, this searches
+    `os.environ['KLAUS_REPOS']` for valid git repositories and updates klaus
+    automatically when any changes are made. if `recursive=True`, klaus will
```

Comment:
This doesn't explain that repos are also looked for in each directory in `KLAUS_REPOS_ROOTS`, i.e. that `KLAUS_REPOS_ROOTS` is a list of directories containing repositories.